runtime/lua/vim/filetype.lua

```diff
@@ -2214,29 +2221,74 @@ local function match_pattern(name, path, tail, pat)
+--- may not always provide a match in cases where the filename (or contents) cannot unambiguously
+--- determine the filetype.
+---
+--- Each of the three options is specified using a key to the single argument of this function.
+--- Example:
+---
+--- <pre>
+---   -- Using a buffer number
+---   vim.filetype.match({ buf = 42 })
 ---
----@param name string File name (can be an absolute or relative path)
----@param bufnr number|nil The buffer to set the filetype for. Defaults to the current buffer.
+---   -- Using a filename
+---   vim.filetype.match({ filename = "main.lua" })
+---
+---   -- Using file contents
+---   vim.filetype.match({ contents = {"#!/usr/bin/env bash"} })
+--- </pre>
+---
+---@param arg table Table specifying which matching strategy to use. It is an error to provide more
+---                 than one strategy. Accepted keys are:
+---                   * buf (number): Buffer number to use for matching
+---                   * filename (string): Filename to use for matching. Note that the file need not
+---                                        actually exist in the filesystem, only the name itself is
+---                                        used.
+---                   * contents (table): An array of lines representing file contents to use for
+---                                       matching.
 ---@return string|nil If a match was found, the matched filetype.
 ---@return function|nil A function that modifies buffer state when called (for example, to set some
 ---                     filetype specific buffer variables). The function accepts a buffer number as
 ---                     its only argument.
-function M.match(name, bufnr)
+function M.match(arg)
   vim.validate({
-    name = { name, 's' },
-    bufnr = { bufnr, 'n', true },
+    arg = { arg, 't' },
   })
 
-  -- When fired from the main filetypedetect autocommand the {bufnr} argument is omitted, so we use
-  -- the current buffer. The {bufnr} argument is provided to allow extensibility in case callers
-  -- wish to perform filetype detection on buffers other than the current one.
-  bufnr = bufnr or api.nvim_get_current_buf()
+  if not (arg.buf or arg.filename or arg.contents) then
+    error('One of "buf", "filename", or "contents" must be given')
+  end
+
+  if (arg.buf and arg.filename) or (arg.buf and arg.contents) or (arg.filename and arg.contents) then
+    error('Only one of "buf", "filename", or "contents" must be given')
+  end
+
+  local bufnr = arg.buf
+  local name = bufnr and api.nvim_buf_get_name(bufnr) or arg.filename
+  local contents = arg.contents
 
-  name = normalize_path(name)
+  if name then
+    name = normalize_path(name)
+  end
 
   local ft, on_detect
 
+  if not (bufnr or name) then
+    -- Sanity check: this should not happen
+    assert(contents, 'contents should be non-nil when bufnr and filename are nil')
+    -- TODO: "scripts.lua" content matching
```

Comment:
In `scripts.lua`, in some cases (like when the first line matches `^csh\>`) the `sh()` function in `detect.lua` is called to retrieve the correct filetype and do additional checks. In this case, won't `bufnr` and `path` be required then?

Reply:
We talked about `pcall`ing the functions if no `bufnr` is provided so we can, in particular, correctly match `filename='eval.cpp'`. We could do the same here?

Reply:
But I agree that this means characterizing `scripts.lua` as "content matching" is slightly misleading -- it is just the last (mostly content-based) step in the matching pipeline, independently of the argument (although it can -- and should -- indeed be skipped for `filename`?)